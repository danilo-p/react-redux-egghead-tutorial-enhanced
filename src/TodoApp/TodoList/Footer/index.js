import React from 'react';
import FilterButton from './FilterButton';
/**
 * Describes the footer of the app.
 *
 * @returns {jsx} The Footer template
 */
const Footer = () => (
  <p>
    Show:&nbsp;&nbsp;
    <FilterButton filter="SHOW_ALL">All</FilterButton>
    <FilterButton filter="SHOW_ACTIVE">Active</FilterButton>
    <FilterButton filter="SHOW_COMPLETED">Completed</FilterButton>
  </p>
);

export default Footer;
