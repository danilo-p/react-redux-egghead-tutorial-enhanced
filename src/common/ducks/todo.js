// Initial State
export const INITIAL_STATE = {};

// Actions
export const ADD = 'ADD_TODO';
export const TOGGLE = 'TOGGLE_TODO';

// Reducer
export default (state = INITIAL_STATE, action) => {
  switch (action.type) {
  case ADD:
    return {
      id: action.id,
      text: action.text,
      completed: false,
    };
  case TOGGLE:
    if (state.id !== action.id) {
      return state;
    }
    return {
      ...state,
      completed: !state.completed,
    };
  default:
    return state;
  }
};

export const createTodoId = text => Date.now() + text;

// Action Creators
export function addTodo(text, id = createTodoId(text)) {
  return {
    type: ADD,
    id,
    text,
  };
}

export const toggleTodo = id => ({
  type: TOGGLE,
  id,
});
