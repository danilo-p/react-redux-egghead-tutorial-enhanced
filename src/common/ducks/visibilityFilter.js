/**
 * The "visibilityFilter" duck.
 *
 * @module common/ducks/visibilityFilter
 */

/* ======================================
=            Initial State              =
====================================== */

/** "visibilityFilter" duck initial state */
export const INITIAL_STATE = 'SHOW_ALL';

/* ======  End of Initial State  ====== */


/* =====================================
=            Action Types              =
===================================== */

/** SET_VISIBILITY_FILTER action type */
export const SET = 'SET_VISIBILITY_FILTER';

/* ======  End of Action Types  ====== */


/* ================================
=            Reducer              =
================================ */

/**
 * The "visibilityFilter" duck reducer
 *
 * @memberof module:common/ducks/visibilityFilter
 * @param {Object} [state=INITIAL_STATE] The current state
 * @param {Object} action The action to be handled
 * @returns {Object} The new state
 */
export default function reducer(state = INITIAL_STATE, action) {
  switch (action.type) {
  case SET:
    return action.filter;
  default:
    return state;
  }
}

/* ======  End of Reducer  ====== */


/* ========================================
=            Action Creators              =
======================================== */

/**
 * Creates a new SET_VISIBILITY_FILTER action.
 *
 * @memberof module:common/ducks/visibilityFilter
 * @param {String} filter The visibility filter
 * @returns {Object} The new SET_VISIBILITY_FILTER action
 */
export const setVisibilityFilter = filter => ({
  type: 'SET_VISIBILITY_FILTER',
  filter,
});

/* ======  End of Action Creators  ====== */
