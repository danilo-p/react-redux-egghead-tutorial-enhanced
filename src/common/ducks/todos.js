import todo, {
  ADD,
  TOGGLE,
} from './todo';

// Actions
export const REMOVE = 'REMOVE_TODO';
export { ADD, TOGGLE };

// Reducer
export default (
  state = [],
  action,
) => {
  switch (action.type) {
  case ADD:
    return [
      ...state,
      todo(undefined, action),
    ];
  case TOGGLE:
    return state.map(t => todo(t, action));
  case REMOVE:
    return []; // TODO: implement the remove todo
  default:
    return state;
  }
};
