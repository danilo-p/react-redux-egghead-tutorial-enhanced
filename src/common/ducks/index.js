/**
 * Module that bundles all the ducks in one reducer. It imports all the ducks
 * and combine their reducers with the combineReducers method of the redux
 * library.
 *
 * @module common/ducks
 */

import { combineReducers } from 'redux';

import todo from './todo';
import todos from './todos';
import visibilityFilter from './visibilityFilter';

// Here all the reducers are combined into a single one reducer that is
// exported as default and used to create the application store.
export default combineReducers({
  todo,
  todos,
  visibilityFilter,
});
