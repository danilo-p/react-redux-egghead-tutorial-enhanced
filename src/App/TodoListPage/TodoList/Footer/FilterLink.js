import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Button } from 'semantic-ui-react';
import { setVisibilityFilter } from 'common/ducks/visibilityFilter';

/**
 * Describes a link that is used to filter the todos on the app.
 *
 * @param {Object} props The component properties
 * @param {Boolean} props.active Is the filter that the link describes active?
 * @param {Array} props.children The link children components
 * @param {Function} props.onClick Button click handler
 * @returns {jsx} The link template
 */
const FilterLink = ({
  active,
  children,
  onClick,
}) => (
  <Button
    secondary
    disabled={active}
    size="mini"
    onClick={onClick}
  >
    {children}
  </Button>
);

FilterLink.propTypes = {
  active: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func.isRequired,
};

const mapStateToProps = (
  state,
  containerProps,
) => ({
  active: containerProps.filter === state.visibilityFilter,
});

const mapDispatchToProps = (
  dispatch,
  containerProps,
) => ({
  onClick: () => {
    dispatch(setVisibilityFilter(containerProps.filter));
  },
});

/**
 * Exports the container component for the FilterLink
 */
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(FilterLink);
