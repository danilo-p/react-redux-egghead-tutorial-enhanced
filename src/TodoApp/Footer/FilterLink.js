import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Button } from 'semantic-ui-react';
import { setVisibilityFilter } from '../ducks/visibilityFilter';

const Link = ({
  active,
  children,
  onClick,
}) => (
  <Button
    secondary
    disabled={active}
    size="mini"
    onClick={onClick}
  >
    {children}
  </Button>
);

Link.propTypes = {
  active: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func.isRequired,
};

const mapStateToProps = (
  state,
  ownProps,
) => ({
  active: ownProps.filter === state.visibilityFilter,
});

const mapDispatchToProps = (
  dispatch,
  ownProps,
) => ({
  onClick: () => {
    dispatch(setVisibilityFilter(ownProps.filter));
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Link);
