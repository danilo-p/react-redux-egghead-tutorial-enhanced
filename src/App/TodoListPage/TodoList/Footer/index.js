import React from 'react';
import FilterLink from './FilterLink';
/**
 * Describes the footer of the app.
 *
 * @returns {jsx} The Footer template
 */
const Footer = () => (
  <p>
    Show:&nbsp;&nbsp;
    <FilterLink filter="SHOW_ALL">All</FilterLink>
    <FilterLink filter="SHOW_ACTIVE">Active</FilterLink>
    <FilterLink filter="SHOW_COMPLETED">Completed</FilterLink>
  </p>
);

export default Footer;
