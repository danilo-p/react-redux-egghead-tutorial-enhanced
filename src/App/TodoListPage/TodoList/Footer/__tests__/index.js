import React from 'react';
import { Provider } from 'react-redux';
import { INITIAL_STATE } from 'common/ducks';
import configureMockStore from 'redux-mock-store';
import Footer from '../';
import FilterLink from '../FilterLink';

const mockStore = configureMockStore();

describe('Footer', () => {
  let footer = null;
  let store = null;

  beforeEach(() => {
    store = mockStore(INITIAL_STATE);

    footer = (
      <Provider store={store}>
        <Footer />
      </Provider>
    );

    footer = mount(footer).find(Footer).first();
  });

  it('should always render one paragraph', () => {
    expect(footer.find('p').length).toBe(1);
  });

  it('should always render exactly 3 FilterLink\'s', () => {
    expect(footer.find(FilterLink).length).toBe(3);
  });

  it('should always render a FilterLink with "filter" property set to' +
    '"SHOW_ALL"', () => {
    expect(footer
      .find(FilterLink)
      .map(fl => fl.props())
      .filter(p => p.filter === 'SHOW_ALL')
      .length).toBe(1);
  });
});
