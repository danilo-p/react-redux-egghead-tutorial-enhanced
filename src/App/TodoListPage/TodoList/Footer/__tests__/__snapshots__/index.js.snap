// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer should render the component 1`] = `
<Footer>
  <p>
    Show:  
    <Connect(FilterLink)
      filter="SHOW_ALL"
    >
      <FilterLink
        active={true}
        filter="SHOW_ALL"
        onClick={[Function]}
      >
        <Button
          as="button"
          disabled={true}
          onClick={[Function]}
          secondary={true}
          size="mini"
        >
          <button
            className="ui mini secondary disabled button"
            disabled={true}
            onClick={[Function]}
            role="button"
            tabIndex={-1}
          >
            All
          </button>
        </Button>
      </FilterLink>
    </Connect(FilterLink)>
    <Connect(FilterLink)
      filter="SHOW_ACTIVE"
    >
      <FilterLink
        active={false}
        filter="SHOW_ACTIVE"
        onClick={[Function]}
      >
        <Button
          as="button"
          disabled={false}
          onClick={[Function]}
          secondary={true}
          size="mini"
        >
          <button
            className="ui mini secondary button"
            disabled={undefined}
            onClick={[Function]}
            role="button"
            tabIndex={undefined}
          >
            Active
          </button>
        </Button>
      </FilterLink>
    </Connect(FilterLink)>
    <Connect(FilterLink)
      filter="SHOW_COMPLETED"
    >
      <FilterLink
        active={false}
        filter="SHOW_COMPLETED"
        onClick={[Function]}
      >
        <Button
          as="button"
          disabled={false}
          onClick={[Function]}
          secondary={true}
          size="mini"
        >
          <button
            className="ui mini secondary button"
            disabled={undefined}
            onClick={[Function]}
            role="button"
            tabIndex={undefined}
          >
            Completed
          </button>
        </Button>
      </FilterLink>
    </Connect(FilterLink)>
  </p>
</Footer>
`;
